### DISCUSSION TITLE
Changes to default BoTorch covariance and likelihood modules

### DISCUSSION THREAD

**Discussion by hvarfner** [meta-pytorch/botorch#2451] [General]
### Dimension-scaled Lengthscale Prior

The conventional Scale-Matern kernel with a $p(\ell) \sim \Gamma(3, 6)$ lengthscale prior yields good results on low-dimensional problems, but becomes increasingly local as the dimensionality of the problem increases due to its emphasis on short lengthscales. To make the out-of-the-box performance of the BoTorch models less susceptible to problem dimensionality, we are replacing the default covariance module with an RBF kernel that utilizes a LogNormal prior: $p(\ell) \sim \mathcal{LN}(\sqrt{2} + \log{\sqrt{D}}, \sqrt{3})$ that scales with the dimensionality of the problem, as proposed by Hvarfner et. al [1].

### Outputscale and Noise Prior Changes
The change also replaces the noise prior of the GaussianLikelihood with a LogNormal prior that prefers lower values, fixes the outputscale (to 1) and switches the default kernel from Matern to RBF, all proposed in [1]. An argument `use_rbf_kernel` has been added to the new helper `get_covar_module_with_dim_scaled_prior` in `models/utils/gpytorch_modules` for users who wish to revert some of the change. The old covariance module is still available through the `get_matern_kernel_with_gamma_prior` helper. Similarly, the new default noise prior is available through `get_gaussian_likelihood_with_lognormal_prior` and the old one through `get_gaussian_likelihood_with_gamma_prior`.

### Planned Further Changes

Since the new covariance module does not include an outputscale parameter, we expect the model performance to be more sensitive to standardization of the training outcomes. In order to alleviate this issue and further improve the reliability of default BoTorch models, we will start utilizing the Normalize and Standardize transforms by default. This change will automate one of the BoTorch best practices and simplify the setup required for utilizing the BoTorch models.

### Supporting Results
In the attached image, the density of the old (blue) and the new priors are displayed for various dimensionalities. The new lengthscale priors are broader, and generally prefer larger values. The new noise prior prefers lower values, but still suggests a moderate level of noise.

In addition to the results in [the paper](https://arxiv.org/abs/2402.02229), we ran a large benchmark study to evaluate the new covariance and likelihood modules against the previous defaults, the [SAASBO](https://arxiv.org/abs/2103.00349) model, and some other variants we utilize internally. We found the new default, and variations of it, to perform near the top across a wide range of benchmark problems, including synthetic test functions and benchmarks built using internal datasets. 

[1] Vanilla Bayesian Optimization Performs Great in High Dimensions. Carl Hvarfner, Erik Orm Hellsten, Luigi Nardi. International Conference on Machine Learning, 2024. URL: https://arxiv.org/abs/2402.02229

![dimscaled_prior](https://github.com/user-attachments/assets/aa4e6674-9a4b-4178-a36f-4cb5469b2741)

**Comment by slishak-PX** [comment]
Hi @hvarfner,

Thank you very much for making this change - I almost always add the LogNormal prior from your paper, along with `Normalize` and `Standardize`. This will substantially streamline my code.

I notice that the HOGP covariance module is unchanged; is this intentional? I think the release notes for 0.12.0 imply that it now uses the dimension-scaled prior.

https://github.com/pytorch/botorch/blob/ad9978fac775e1657bb4e5375d85e443913da25e/botorch/models/higher_order_gp.py#L252-L257

**Reply by saitcakmak** [reply]
Hi @slishak-PX. I had ran into some test failures while updating the covar module for HOGP and decided to skip it at the time. I'll give it another try.

**Reply by saitcakmak** [reply]
https://github.com/pytorch/botorch/pull/2555 updates it.

**Comment by slishak-PX** [comment]
I'm finding in some cases that I get better performance with `use_rbf_kernel=False`, but it is slightly annoying having to call `get_covar_module_with_dim_scaled_prior` with the correct `ard_num_dims` and `batch_shape` and pass that in as `covar_module`. It would be nice if there was a way to pass the `use_rbf_kernel` argument down from the model constructor.

I appreciate that this would pollute the API a bit, as a lot of models would now have an extra `use_rbf_kernel` argument that is ignored if `covar_module` is also set, so I'm just leaving this here in case other people have better ideas.

**Comment by theo-brown** [comment]
Somewhat late to the party here, but great to see that this has been integrated into BoTorch.
@hvarfner following from the comment above, are there empirical demos of where one might choose RBF vs Matern if using the prior and likelihood from your paper?

**Comment by hvarfner** [comment]
Hi @theo-brown ,

Are you referring to how you can make the choice of using Matern vs. RBF when creating a model?

**Reply by theo-brown** [reply]
Yes - in your original comment in this discussion you mentioned that following your work the default kernel in Botorch has been changed from Matern to RBF, and I was wondering why that was. 

While your paper provided excellent justification for using a dimensionality scaled prior on the lengthscale and a fixed signal outputscale, I couldn't see much detail on RBF vs Matern even for the standard synthetic functions.

I'm just a bit surprised by changing the default to RBF, because I was under the impression that the added roughness of the Matern kernels was generally regarded as useful for real-world BO problems. Maybe that's another myth you've busted!

**Comment by hvarfner** [comment]
Ah, I see! Thanks for the kind words, and I agree that the motivation of RBF/Matern is not very well justified.

For the experiments in the paper, I found that RBF, on aggregate, outperformed Matern pretty much across the board. For some reason, those did not make it into the paper - I'll see if I still have them. 

This was ablated on a set of internal baselines as well, and the results were consistent with what I had previously found. Now, these tasks were mostly high-dimensional. 

Now, this could change could absolutely have been better supported with some tangible, public results, so I'll see if I can dig some up. Nonetheless, it is _not_ a make-or-break switch, so if you prefer Matern, I would absolutely say go for it!

Here are the change in code.

```
import torch
from botorch.models import SingleTaskGP
from botorch.models.utils.gpytorch_modules import get_covar_module_with_dim_scaled_prior


train_X = torch.rand(10, 3, dtype=torch.float64)  # 10 points in 3D
train_Y = torch.rand(10, 1, dtype=torch.float64)

gp_model = SingleTaskGP(
    train_X=train_X,
    train_Y=train_Y,
    covar_module=get_covar_module_with_dim_scaled_prior(
        ard_num_dims=train_X.shape[-1],
        use_rbf_kernel=False, # HERE IT IS =)
    )
)
```

**Reply by theo-brown** [reply]
Fair enough! Thanks for trying to dig out some more public tangible results on this :)
If the superiority of RBF over Matern breaks down in lower dimensional tasks it might be good to have some sort of note on this in the docs.

**Reply by bchen0** [reply]
Hi @hvarfner - just to check, it seems that setting covar_module as None also has a few additional effects. Would you be able to clarify what these self._subset_batch_dict settings do? I checked Model.subset_output but did not get very far.

It seems that even setting covar_module with get_covar_module_with_dim_scaled_prior with use_rbf_kernel=True in your above code would not be exactly the same as setting covar_module=None.

In gp_regression.SingleTaskGP:
```
if covar_module is None:
    covar_module = get_covar_module_with_dim_scaled_prior(
        ard_num_dims=transformed_X.shape[-1],
        batch_shape=self._aug_batch_shape,
    )
    # Used for subsetting along the output dimension. See Model.subset_output.
    self._subset_batch_dict = {
        "mean_module.raw_constant": -1,
        "covar_module.raw_lengthscale": -3,
    }
    if train_Yvar is None:
        self._subset_batch_dict["likelihood.noise_covar.raw_noise"] = -2
self.covar_module: Module = covar_module
```
